nupic/research/frameworks/pytorch/imagenet/imagenet_experiment.py

```diff
@@ -869,47 +894,173 @@ def get_free_port(self):
+            transform_model=[exp + ".transform_model"],
+            create_loaders=[exp + ".create_loaders"],
+            create_train_dataloader=[exp + ".create_train_dataloader"],
+            create_train_sampler=[exp + ".create_train_sampler"],
+            create_validation_dataloader=[exp + ".create_validation_dataloader"],
+            create_validation_sampler=[exp + ".create_validation_sampler"],
+            create_lr_scheduler=[exp + ".create_lr_scheduler"],
+            pre_experiment=[exp + ".pre_experiment"],
+            validate=[exp + ".validate"],
+            train_epoch=[exp + ".train_epoch"],
+            run_epoch=[exp + ".run_epoch"],
+            pre_epoch=[exp + ".pre_epoch"],
+            post_epoch=[exp + ".post_epoch"],
+            pre_batch=[exp + ".pre_batch"],
+            post_batch=[exp + ".post_batch"],
+            error_loss=[exp + ".error_loss"],
+            complexity_loss=[exp + ".complexity_loss"],
+            should_decay_parameter=[exp + ".should_decay_parameter"],
+            transform_data_to_device=[exp + ".transform_data_to_device"],
+            aggregate_results=[exp + ".aggregate_results"],
+            aggregate_validation_results=[exp + ".aggregate_validation_results"],
             aggregate_pre_experiment_results=[
-                "SupervisedExperiment.aggregate_pre_experiment_results"
-            ],
-            get_printable_result=["SupervisedExperiment.get_printable_result"],
-            expand_result_to_time_series=[
-                "SupervisedExperiment: validation results"
+                exp + ".aggregate_pre_experiment_results"
             ],
-            stop_experiment=["SupervisedExperiment.stop_experiment"]
+            get_printable_result=[exp + ".get_printable_result"],
+            expand_result_to_time_series=[exp + ": validation results"],
+            stop_experiment=[exp + ".stop_experiment"]
         )
 
 
+class ContinualLearningExperiment(ContinualLearningMetrics, SupervisedExperiment):
+
+    def setup_experiment(self, config):
+
+        super().setup_experiment(config)
+        # override epochs to validate to not validate within the inner loop over epochs
+        self.epochs_to_validate = []
+
+        self.current_task = 0
+
+        # defines how many classes should exist per task
+        self.num_tasks = config.get("num_tasks", 1)
+
+        self.num_classes = config.get("num_classes", None)
+        assert self.num_classes is not None, "num_classes should be defined"
+
+        self.num_classes_per_task = math.floor(self.num_classes / self.num_tasks)
+
+        # applying target transform depending on type of CL task
+        # task - we know the task, so the network is multihead
+        # class - we don't know the task
+        self.cl_experiment_type = config.get("cl_experiment_type", "class")
+        if self.cl_experiment_type == "task":
+            # override the target transform
+            self.dataset_args["target_transform"] = (
```

Comment:
Would this need to be passed to the validation dataset as well? I don't think this happens as of now.

Also, there's a lurking error where the code now assumes "target_transform" is accepted to your dataset constructor. A more general way to do this may be 
1. create your datasets
2. assert they are a VisionDataset if `cl_experiment_type == "task"`
3. and then modify their `target_transform`
That's a bit sly though. 

You could also possibly throw a warning when you modify the `dataset_args`, just to make sure the user knows the argument is being overridden and that it might not work with the `dataset_class` they've given.

Reply:
`dataset_args` applies equally to both training and validation datasets (which is probably an issue, we should find a way of setting different transforms and target_transforms to both). 

Good catch on the error, I will work on the remaining changes. Truth is we never really tested the `task` approach until now, only the `class` - these will come up more obviously once we implement EWC and start comparing results.